### PR TITLE
repo: add github actions CI

### PR DESCRIPTION
**Pull request by WanzenBug**
Create infrastructure for https://pre-commit.com/ and checking via
github actions.

### CODE REVIEW

**Comment by JoelColledge** [comment]
Interesting approach. Do you have experience with `pre-commit`?

**Comment by WanzenBug** [comment]
I've used it previously for some Python projects. It makes it quite easy to integraten"standard" steps like running `go test` with steps like recreating crds from models, fixing whitespace etc.

**Comment by WanzenBug** [comment]
The infrastructure is almost there now. However, since we haven't been running golangci-lint for some time, there are a lot of changes to make. I would wait with fixing them until after the linstor-client HTTPS feature is implemented. 

I used my [personal fork to test the CI setup](https://github.com/WanzenBug/piraeus-operator/pull/2) (which reports a lot more files as changed, so I get a lot of linter warnings) 
image

**Comment by WanzenBug** [comment]
This is ready for review. I did fix all the issues reported by golangci-lint and gofumpt, they can be fixed as they come up when changing files.

Checks are all green now: https://github.com/WanzenBug/piraeus-operator/pull/2/checks

**Comment by WanzenBug** [comment]
@JoelColledge could you please take a look?

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+          key: ${{ runner.os }}-build-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-build-
+      - name: Build with operator-sdk
+        run: operator-sdk build quay.io/piraeusdatastore/piraeus-operator:latest
+      - name: Push docker image
+        uses: docker/build-push-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
```
Comment (JoelColledge) [inline .github/workflows/build.yml:40]: Do these secrets already exist?

Comment (WanzenBug) [reply .github/workflows/build.yml:40]: No :sweat_smile:
Automatic push is the only thing I could not really test in my fork (docker build works fine)

Comment (JoelColledge) [reply .github/workflows/build.yml:40]: @rck Could you configure secrets for pushing to `quay.io`? I don't have the permissions.

File: .github/workflows/golangci.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+jobs:
+  golangci:
+    name: lint
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v1
+        with:
+          # Required: the version of golangci-lint is required and must be specified without patch version: we always use the latest patch version.
```
Comment (JoelColledge) [inline .github/workflows/golangci.yml:18]: No need to copy the documentation of the action in here

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  build:
+    name: build
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-go@v2.0.3
+      - uses: actions/cache@v1
+        with:
+          path: ~/operator-sdk/bin/
+          key: operator-sdk-${{ env.OPERATOR_SDK_VERSION }}
```
Comment (JoelColledge) [inline .github/workflows/build.yml:19]: Where does `OPERATOR_SDK_VERSION` actually come from? I see that it is set explicitly in `pre-commit.yml`, but what about here?

File: .github/workflows/golangci.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+name: golangci-lint
```
Comment (JoelColledge) [inline .github/workflows/golangci.yml:1]: Can we not run this as part of precommit?

Comment (WanzenBug) [reply .github/workflows/golangci.yml:1]: We can, the advantage of this approach is that the github action will automatically create notes on the relevant diffs

File: .github/workflows/build.yml (hunk 3 above)
Comment (WanzenBug) [reply .github/workflows/build.yml:19]: copy&paste error

File: .github/workflows/golangci.yml (hunk 4 above)
Comment (JoelColledge) [reply .github/workflows/golangci.yml:1]: OK, that makes it worthwhile

File: .github/workflows/golangci.yml (hunk 2 above)
Comment (WanzenBug) [reply .github/workflows/golangci.yml]: Removed

File: .github/workflows/build.yml (hunk 3 above)
Comment (WanzenBug) [reply .github/workflows/build.yml:21]: fixed

**Comment by JoelColledge** [comment]
Any idea why the checks aren't running for this PR?

**Comment by WanzenBug** [comment]
I think because I'm "external" for the purpose of this repo, so I can't create any workflows.

**Comment by WanzenBug** [comment]
In my fork they are running as expected: https://github.com/WanzenBug/piraeus-operator/pull/2/checks

File: charts/piraeus/crds/piraeus.linbit.com_linstorcontrollersets_crd.yaml [hunk 5]
```diff
@@ -32,47 +32,33 @@ spec:
...
-                controller/server container
-              type: string
-            linstorHttpsControllerSecret:
-              description: Name of the secret containing the java keystore (`keystore.jks`)
-                used to enable HTTPS on the controller. The controller will create
-                a secured https endpoint on port 3371 with the key stored in `keystore.jks`.
-                The keystore must be secured using the passphrase "linstor". Also needs to
-                contain a truststore `truststore.jks`, which will be used to authenticate
-                clients.
+              description: DBCertSecret is the name of the kubernetes secret that
```
Comment (JoelColledge) [inline charts/piraeus/crds/piraeus.linbit.com_linstorcontrollersets_crd.yaml:57]: This description doesn't belong with the key `drbdRepoCred`.

I will stop reviewing here and let you reconcile the differences between the CRDs are they were and these newly generated ones. This yaml differ https://eloquent-hodgkin-f52b42.netlify.app/ seems to work pretty well. The result doesn't have to be identical to the original, but we should make sure we're not making anything worse.

File: .github/workflows/build.yml (hunk 1 above)
Comment (rck) [reply .github/workflows/build.yml:42]: I will take care of that.

File: charts/piraeus/crds/piraeus.linbit.com_linstorcontrollersets_crd.yaml (hunk 5 above)
Comment (WanzenBug) [reply charts/piraeus/crds/piraeus.linbit.com_linstorcontrollersets_crd.yaml]: fixed, result is identical for validation purpose. Updated the descriptions with whatever seemed more appropriate from the old CRDs or go comments

File: .github/workflows/build.yml (hunk 1 above)
Comment (JoelColledge) [reply .github/workflows/build.yml:42]: The secrets now exist

**Pull request merged**